### PR TITLE
windows: Move IMMDevice work to common file, implement DirectSound enumeration support

### LINKED ISSUE #5750: Get the DirectSound device spec using IMMDevice, when available
Right now DirectSound doesn't provide the endpoint format, because it didn't provide that information until the context was fully made. However, with Vista and newer, _supposedly_ it's possible to take the DSound endpoints at enumeration time and link them to an IMMDevice, at which point which we can get the device format the same way we get it for WASAPI. This should allow DirectSound to support SDL_GetAudioDeviceSpec.

### PR DESCRIPTION
**Pull request by flibitijibibo**
This is a draft, because:

- [x] directsound.c changes are a massive copypaste from wasapi.c, working on this now
- [x] Need to test for consistency between the old and new enumerators
- [x] Update GDK project
- [x] Fix WinRT build (and Watcom)

Other than that, this was actually really easy! We're mostly looking at cleanup, but still wanted to get Ryan's eyes on this as soon as I had something building.

Fixes #5750

### CODE REVIEW

**Comment by sezero** [comment]
`$srcdir/src/core/windows/*.c` is added to sources list which will include the new SDL_immdevice.c:
HAVE_MMDEVICEAPI_H guarding for  must be (also / instead ?) done in SDL_immdevice.[ch]

File: src/core/windows/SDL_immdevice.c [hunk 1]
```diff
@@ -20,7 +20,7 @@
 */
 #include "../../SDL_internal.h"
 
-#if defined(__WIN32__) || defined(__GDK__)
+#if (defined(__WIN32__) || defined(__GDK__)) && HAVE_MMDEVICEAPI_H
```
Comment (sezero) [inline src/core/windows/SDL_immdevice.c:23]: You can now add SDL_immdevice.c to Makefile.w32 for completeness:
```diff
diff --git a/Makefile.w32 b/Makefile.w32
index 1ca1fa2..978afab 100644
--- a/Makefile.w32
+++ b/Makefile.w32
@@ -76 +76 @@
-SRCS+= SDL_hid.c SDL_windows.c SDL_xinput.c
+SRCS+= SDL_hid.c SDL_immdevice.c SDL_windows.c SDL_xinput.c
```

PS: missing newline at end of SDL_immdevice.[ch]

**Comment by flibitijibibo** [comment]
Watcom is working again, UWP is being a pain... going to dust off a UWP build system and work on this locally. Will also add the newlines to SDL_immdevice in the meantime.

**Comment by flibitijibibo** [comment]
UGH OKAY WinRT finally cooperated. To keep it stable it's just preserving the original enumeration logic, since it has a whole other system for device enumeration that annoyingly doesn't use IMMDevice.

This should behave pretty much identically to before on every platform except DirectSound, since everything was copypasted quite aggressively. The one thing we will want to test is enumeration; as long as the device list isn't wildly different this should be good to go; the backend still uses GUIDs so all the existing code will continue to work as if we used the old DSound enumeration APIs.

**Comment by flibitijibibo** [comment]
Tested on Windows 7 and confirmed the audio device names are the same, and that using the new path means SDL_GetAudioDeviceSpec works with directsound! This just needs one more +1 for sanity.

**Comment by flibitijibibo** [comment]
Fixed a memory leak last night to address the last of what I could find in local tests. I looked at the milestone and it's exactly two weeks from today, and we haven't pushed a pre-release in about three weeks, so I'm going to try the Organ Donor approach: If nobody says _no_ to this patch I'll merge tomorrow morning, since having a way around the WASAPI resampling back-and-forth without compromising features is critical for a handful of high-profile apps (TMNT is the extra big one).

**Comment by slouken** [comment]
Yes please. Feel free to merge this when you're comfortable with it.

**Pull request merged**